docs/js-core-2/cowsay-project/readme.md

```diff
@@ -0,0 +1,120 @@
+```
+mkdir /cowsaying
+cd /cowsaying
+npm init
+```
+Press return through all the options and then do.
+```
+npm install cowsay
+```
+### npm run
+
+Let's find out how to execute cowsay from our package.json. Open /cowsaying in your text editor and look at the [package.json](https://nodejs.dev/learn/the-package-json-guide).
+
+Add cowsay to the scripts array. 
+```
+"scripts": {
+    "cowsay": "cowsay Mooooo"
+  },
+```
+Now cowsay is listed as a script, and it has an argument.  In your terminal, run cowsay.
+```
+npm run cowsay
+```
+What are the downsides of this approach? I would say one is that we want to be able to pass in arguments, not just always say moo. How could we make the cow say Baaaa?
+
+Remove the hardcoded "Mooooo".  Can you pass a different string in the command line?
+
+### node 
+
+Now let's execute cowsay in a different way. Open /index.js in your text editor. Import the cowsay package with require.
+```
+const cowsay = require("cowsay");
+```
+And log a cow to the console that says "I require a package". Add some more arguments if you can.
+```
+console.log(cowsay.say({
+        text : "I require a package",
+        e : "oO",
+        T : "U "
+    }));
+```
+Now run this with node:
+```
+node index.js
+```
+## Going deeper, your stretch goal
+
+How is this cow being made and what else can we make? Open your node modules and look inside cowsay. There's a README file that tells you all about the package and a library of other animals you can make. Read it. You don't have to understand all the code to get an idea of what is going on. 
+
+How would you extend this library to add a new animal? What about [a duck](https://www.asciiart.eu/animals/birds-water)? Look at how the existing ```.cows``` are imported and try adapting this for your own .cow. There are no solutions provided for this challenge. Dig around for yourself.
+
+## Project
+
+Let's figure out how to make a cow say things in Node by ourselves. For this project we don't really need a package, a library or lots of options. Let's just get the cow printing out and saying whatever we write in the command line. What would be helpful? I think we need to:
+
+* [Accept an argument](https://nodejs.dev/learn/nodejs-accept-arguments-from-the-command-line) from the command line.
+* Output to the command line. You've already done this with console.log.
+* Make an ASCII cow.
+* Write a function that puts the string into the cow's speech bubble.
+
+Write your solution in a file called solution.js and test it by running your program in the command line. How will you handle a null argument?
+
+### Iterating
+
+We could make our program more accessible by adding a command line interface that prompts us to write in the cow's words. What tools can we use? I think we could use:
+
+* A command line interface. I'll start you off by letting you know that there is a built in CLI called [readline](https://nodejs.dev/learn/accept-input-from-the-command-line-in-nodejs).
```

Comment:
I think writing async callback functions is sadly a bit advanced to just link to for JS1/JS2 and expect the trainees to be able to follow - if we're going to use `readline`, we'll probably need a bit more scaffolding or a step-by-step to follow... Could be really interesting though, but should probably be marked pretty clearly as a very optional extension

Reply:
Agreed - I think with some scaffolding around how to accept command line argument it could be a really nice exercise

Reply:
Yes, I agree with this. What about splitting cowsay into two: Part one is just investigating installations and running things in terminal and  then return to it in JS3 where we write our own cowsay and iterate on that. Returning to cowsay and being able to write your own by JS3 would create a satisfying experience of progress and mastery.

Reply:
...That's with some more scaffolding around the first cowsay and then this time using it to layer up on features - could maybe do readline, commander, and then a library of cows?

Reply:
I like the idea of splitting up the project, however I would point out that we don't actually touch NPM at all until React in the current syllabus. I'm actually not quite sure at what point we introduce the terminal now? My feeling is that we should really focus on the basics of JS before introducing packages.

I was actually going to suggest that the first section of the project might be useful as a lesson/coursework in JS3 as an fun introduction to packages.

The second half would be useful as a challenge, especially if we gave some more step-by-step instructions or provided some scaffolding.

Reply:
Yeah, pushing the whole thing into JS3 sounds great!